### PR TITLE
[SPARK-11612] [ML] Pipeline and PipelineModel persistence

### PR DESCRIPTION
**Pull request by jkbradley**
Pipeline and PipelineModel extend Readable and Writable.  Persistence succeeds only when all stages are Writable.

Note: This PR reinstates tests for other read/write functionality.  It should probably not get merged until [https://issues.apache.org/jira/browse/SPARK-11672] gets fixed.

CC: @mengxr

### CODE REVIEW

**Comment by mengxr** [comment]
test this please

File: mllib/src/main/scala/org/apache/spark/ml/Pipeline.scala [hunk 1]
```diff
@@ -166,6 +173,34 @@ class Pipeline(override val uid: String) extends Estimator[PipelineModel] {
...
+object Pipeline extends Readable[Pipeline] {
+
+  override def read: Reader[Pipeline] = new PipelineReader
+
+  override def load(path: String): Pipeline = read.load(path)
+}
+
+private[ml] class PipelineWriter(instance: Pipeline) extends Writer {
+
+  PipelineSharedWriter.validateStages(instance.getStages)
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/Pipeline.scala]: Should users be able to save an incomplete pipeline? For example, I could make a template pipeline, send it to other users, and they only need to fill in some required params like inputCol after they load it back.

File: mllib/src/main/scala/org/apache/spark/ml/Pipeline.scala [hunk 2]
```diff
@@ -200,4 +235,121 @@ class PipelineModel private[ml] (
...
+
+  /** Checked against metadata when loading model */
+  private val className = "org.apache.spark.ml.PipelineModel"
+
+  override def load(path: String): PipelineModel = {
+    val (uid: String, stages: Array[PipelineStage]) =
+      PipelineSharedReader.load(className, sc, path)
+    val transformers = stages map {
+      case stage: Transformer => stage
+      case stage => throw new RuntimeException(s"PipelineModel.read loaded a stage but found it" +
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/Pipeline.scala]: minor: `stage` -> `other`

File: mllib/src/main/scala/org/apache/spark/ml/Pipeline.scala [hunk 3]
```diff
@@ -200,4 +235,121 @@ class PipelineModel private[ml] (
...
+        s" was not a Transformer.  Bad stage: ${stage.uid}")
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/Pipeline.scala]: include `stage.getClass` in the error message

File: mllib/src/main/scala/org/apache/spark/ml/Pipeline.scala [hunk 4]
```diff
@@ -200,4 +235,121 @@ class PipelineModel private[ml] (
...
+/** Methods for [[Writer]] shared between [[Pipeline]] and [[PipelineModel]] */
+private[ml] object PipelineSharedWriter {
+
+  import org.json4s.JsonDSL._
+
+  /** Check that all stages are Writable */
+  def validateStages(stages: Array[PipelineStage]): Unit = {
+    stages.foreach {
+      case stage: Writable => // good
+      case stage =>
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/Pipeline.scala]: minor: `other`

File: mllib/src/main/scala/org/apache/spark/ml/Pipeline.scala [hunk 5]
```diff
@@ -200,4 +235,121 @@ class PipelineModel private[ml] (
...
+        throw new UnsupportedOperationException("Pipeline write will fail on this Pipeline" +
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/Pipeline.scala]: minor: `will fail` -> `failed`, `this Pipeline` -> `pipeline ${pipeline.uid}`

File: mllib/src/main/scala/org/apache/spark/ml/Pipeline.scala [hunk 6]
```diff
@@ -200,4 +235,121 @@ class PipelineModel private[ml] (
...
+          s" because it contains a stage which does not implement Writable. Non-Writable stage:" +
+          s" ${stage.uid}")
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/Pipeline.scala]: Include `stage.getClass`. Though the default implementation of `uid` contains class name, user can set arbitrary values.

File: mllib/src/main/scala/org/apache/spark/ml/Pipeline.scala [hunk 7]
```diff
@@ -200,4 +235,121 @@ class PipelineModel private[ml] (
...
+    }
+  }
+
+  def saveImpl(
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/Pipeline.scala]: minor: `saveImpl` -> `save`?

File: mllib/src/main/scala/org/apache/spark/ml/Pipeline.scala [hunk 8]
```diff
@@ -200,4 +235,121 @@ class PipelineModel private[ml] (
...
+      instance: Params,
+      stages: Array[PipelineStage],
+      sc: SparkContext,
+      path: String): Unit = {
+    // Copied and edited from DefaultParamsWriter.saveMetadata
+    // TODO: modify DefaultParamsWriter.saveMetadata to avoid duplication
+    val uid = instance.uid
+    val cls = instance.getClass.getName
+    val stageUids = stages.map(_.uid)
+    val jsonParams = List("stageUids" -> parse(compact(render(stageUids.toSeq))))
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/Pipeline.scala]: Pass `stageUids` directly if you define `stageUids` as `stages.map(_.uid).toSeq`. I guess it should work.

File: mllib/src/main/scala/org/apache/spark/ml/Pipeline.scala [hunk 9]
```diff
@@ -200,4 +235,121 @@ class PipelineModel private[ml] (
...
+      ("paramMap" -> jsonParams)
+    val metadataPath = new Path(path, "metadata").toString
+    val metadataJson = compact(render(metadata))
+    sc.parallelize(Seq(metadataJson), 1).saveAsTextFile(metadataPath)
+
+    // Save stages
+    val stagesDir = new Path(path, "stages").toString
+    stages.foreach {
+      case stage: Writable =>
+        val stagePath = new Path(stagesDir, stage.uid).toString
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/Pipeline.scala]: It is useful if we prefix the stagePath with index, e.g., `0_Tokeninzer_123/, 1_LogisticRegression_456/`. This helps manual inspection.

File: mllib/src/main/scala/org/apache/spark/ml/Pipeline.scala [hunk 10]
```diff
@@ -200,4 +235,121 @@ class PipelineModel private[ml] (
...
+    val stageUids: Array[String] = metadata.params match {
+      case JObject(pairs) =>
+        if (pairs.length != 1) {
+          // Should not happen unless file is corrupted or we have a bug.
+          throw new RuntimeException(
+            s"Pipeline read expected 1 Param (stageUids), but found ${pairs.length}.")
+        }
+        pairs.head match {
+          case ("stageUids", jsonValue) =>
+            parse(compact(render(jsonValue))).extract[Seq[String]].toArray
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/Pipeline.scala]: Would `jsonValue.extract[Seq[String]].toArray` work?

File: mllib/src/main/scala/org/apache/spark/ml/Pipeline.scala [hunk 11]
```diff
@@ -200,4 +235,121 @@ class PipelineModel private[ml] (
...
+        }
+      case _ =>
+        throw new IllegalArgumentException(
+          s"Cannot recognize JSON metadata: ${metadata.metadataStr}.")
+    }
+    val stages: Array[PipelineStage] = stageUids.map { stageUid =>
+      val stagePath = new Path(stagesDir, stageUid).toString
+      val stageMetadata = DefaultParamsReader.loadMetadata(stagePath, sc)
+      val cls = Utils.classForName(stageMetadata.className)
+      cls.getMethod("read").invoke(cls).asInstanceOf[Reader[PipelineStage]].load(stagePath)
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/Pipeline.scala]: `invoke(null)`

**Comment by mengxr** [comment]
One suggestion is to merge `PipelineShardWriter` and `PipelineSharedReader` into a single object under `object Pipeline`, e.g., called `SharedReadWrite`. Then move `PipelineReader`, `PipelineWriter` to `object Pipeline`, and `PipelineModelReader` and `PipelineModelWriter` to `object PipelineModel`. The main purpose it to not pollute the package space in Java. Otherwise, they are all visible under `org.apache.spark.ml` in Java.

File: mllib/src/main/scala/org/apache/spark/ml/Pipeline.scala (hunk 5 above)
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/Pipeline.scala]: But a user could write:
```
val writer = pipeline.write   // failure will occur here, before attempting to write
writer.save(...)
```

File: mllib/src/main/scala/org/apache/spark/ml/Pipeline.scala (hunk 8 above)
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/Pipeline.scala]: Is another way than this better?

File: mllib/src/main/scala/org/apache/spark/ml/Pipeline.scala (hunk 7 above)
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/Pipeline.scala]: I like saveImpl since it's analogous to the other saveImpl methods which call it

**Comment by jkbradley** [comment]
@mengxr  Thanks for reviewing!  I believe I addressed everything, except where I quibbled in responses above.

**Comment by mengxr** [comment]
LGTM pending Jenkins.

**Comment by jkbradley** [comment]
@mengxr Thank you for reviewing!  Merging with master and branch-1.6

**Pull request closed without merging**